bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/ScriptEngineManagerImpl.java

```diff
@@ -163,6 +169,22 @@ public void loadScript(String engineIdentifier, InputStreamReader scriptData) {
         }
     }
 
+    private void tryExtractAndInjectFilename(String engineIdentifier, ScriptEngine engine) {
+        try {
+            if(engine.getContext().getAttribute(ScriptEngine.FILENAME) == null) {
+                URI fileURI = new URL(engineIdentifier).toURI();
+                if ("file".equals(fileURI.getScheme())) {
+                    File maybeScriptFile = Paths.get(fileURI).toFile();
+                    if (maybeScriptFile.isFile()) {
+                        engine.getContext().setAttribute(ScriptEngine.FILENAME, maybeScriptFile.getName(), ScriptContext.ENGINE_SCOPE);
```

Comment:
Does you know how this looks in a stack? Here is a sample stack trace:

```
org.graalvm.polyglot.PolyglotException: Item 'LivingRoomLights_Light_Real' could not be found in the item registry
        at org.eclipse.smarthome.core.internal.items.ItemRegistryImpl.getItem(ItemRegistryImpl.java:82) ~[?:?]
        at <js>.getItem(items.js:347) ~[?:?]
        at <js>.:program(color_lights.js:12) ~[?:?]
        at org.graalvm.polyglot.Context.eval(Context.java:344) ~[?:?]
        <snip>
```
You can see the script part of it in the middle, with each frame starting with `<js>`. One frame was auto-filled by the engine, because it loaded the file (items.js), the other was populated by the code in this PR because it came from a stream as it was the entry-point script loaded by OH (color_lights.js).
I am asking whether you have seen what happens if you use the full path, because for the other frames, the engine only uses the name.

Reply:
Using getPath(), before...
```
2019-11-24 22:42:12.190 [ERROR] [org.openhab.core.automation.module.script.internal.ScriptEngineManagerImpl] - Error during evaluation of script 'file:/opt/openhab2/conf/automation/jsr223/python/personal/test/test_script_2.py': NameError: name 'PersistenceExtensions' is not defined in <script> at line number 5
```
After...
```
2019-12-07 11:53:44.925 [ERROR] [org.openhab.core.automation.module.script.internal.ScriptEngineManagerImpl] - Error during evaluation of script 'file:/opt/openhab2/conf/automation/jsr223/python/personal/test/test_script_2.py': NameError: name 'PersistenceExtensions' is not defined in /opt/openhab2/conf/automation/jsr223/python/personal/test/test_script_2.py at line number 5
```

Reply:
Interesting! What is producing the Python traceback? The Python engine? If you look at the example I gave, it only includes the filename in both the JS and Java frames. Digging into where these come from, the JS comes from my module loader (the javadoc specified '_the name of the source_'), and the other frames come from the Java runtime (which also only includes the filename). Additionally the script engine param Javadoc specifies that it should be '_the name of the file being executed_'.

Personally I don't think that usability matters too much here; whilst it's possible to have non-unique filenames, in practice you are never working on two files with the same name. I do however see that it's going to look weird to have a mixture of paths & names in either traceback.

One option here would be to pass the engineIdentifier to the ScriptEngineFactory and let it set it, this way we can choose different options for each language, depending on whether their standard formatting uses paths or names?

(As an aside, two things that I have missed and needed to workaround in the factory is (a) never seeing the engineIdentifier and (b) not being notified about disposal of engines.)

Reply:
> whilst it's possible to have non-unique filenames, in practice you are never working on two files with the same name. 

While not extremely common, it is very possible to have scripts of the same name in different directories.

> One option here would be to pass the engineIdentifier to the ScriptEngineFactory and let it set it

I don't see this as an option, since not everyone will be using custom ScriptEngineFactories. Groovy, Jython, jRuby, Kotlin, etc. can just be dropped into the classpath.

Reply:
Sure, I'm not saying that it's impossible, just that in practice it's rare enough that most language implementations don't print the full path. For the languages that I know that run on the JVM, only Python & Ruby print the full path. The others (Java, Kotlin, Groovy, Scala, Javascript, plus Xtend I guess) do not, and nor do languages like C & C++. (Personally I find the full path just means you get long, confusing frames, but I accept others may not.) Additionally the docs point to using the name rather than the path.

I will switch to path if you really think it's the right way to go, but it feels like an optimisation for Jython here, rather than 'most' languages.

As for letting the ScriptEngineFactory set it; maybe we would set it if they had not? So for any language that does _not_ set it, use the most common (name), but if it's already been set (as for example Jython is in it's own bundle, so could set it as the path), then we leave it? This means that we get the expected/idiomatic behaviour for all languages except JRuby?

Reply:
> Additionally the docs point to using the name rather than the path.

Could you please provide a link? I cannot find any reference that states the path should not be used? Still, it's up to us to decide which is more functional for OH users.

> I will switch to path if you really think it's the right way to go

:+1: :stuck_out_tongue_winking_eye: 

> but it feels like an optimisation for Jython here, rather than 'most' languages.

How so? All of the supported scripting languages can have non-unique names for scripts. I don't see anything Jython specific in a solution to provide a unique path rather than a file name to help troubleshoot an exception. Stacktraces are ugly to begin with and I don't think a full path makes it any uglier. Isn't this all irrelevant though, since stacktraces are neutered for automation?

> As for letting the ScriptEngineFactory set it; maybe we would set it if they had not? So for any language that does not set it, use the most common (name), but if it's already been set (as for example Jython is in it's own bundle, so could set it as the path), then we leave it?

Jython does not set this, and there will be people who chose not to use the bundle, so it wold end up being <script>. How about defaulting to full path and having configuration options in the custom ScriptEngineFactories?

Reply:
Ok, links are:
javax.script: https://docs.oracle.com/javase/7/docs/api/javax/script/ScriptEngine.html#FILENAME
graal (examples and API calls): https://www.graalvm.org/sdk/javadoc/org/graalvm/polyglot/Source.html
^ these don't specify that you _shouldn't_ use a path, they just say that you should use a name, and that is what the example uses too.

I say it's an optimisation for Jython because when the stack trace is dumped, it will look wrong for all languages except Jython (and JRuby). For JS it will look like this:
```
        at <js>.myScript(/here/is/where/the/full/path/to/the/script/will/be/file.js:11) ~[?:?]
        at <js>.myFunction(file1.js:347) ~[?:?]
        at <js>.:myOtherFunction(file2.js:12) ~[?:?]
        at <js>.:myFinalFunction(file3.js:12) ~[?:?]
```
While you are correct that currently we have no stack traces thrown from Java code, they can be from script code, so we'll still see this.

I'm not too fussed about which way the default goes and how we switch, but I did have an idea that may solve many of these challenges:
- Add parameter of `engineIdentifier` to `ScriptEngineFactory.createScriptEngine(..)`
- Add parameter of `ScriptExtensionAccessor` (an interface extracted from `ScriptExtensionManager` that provides readonly access) to the same method

This way, in this case, the ScriptEngineFactories can set the filename (the generic one could choose the default). It would also mean that they get some control over how presets are accessed, so would remove the need I had for the other PR about the scriptExtension.importPreset call. Happy to submit another PR if you think it would be the way to go.

Reply:
I've tested this PR for a while now and I have not seen any issues. Having the script path in exceptions is helpful. I still feel the path should be used rather than the filename, since script file names are not unique. This is true for all languages.

If I were a maintainer here, I'd hold out on merging until this was corrected.